# Mutate-Compute

This suite of Python classes and Python scripts is designed to run primarily on the HPC OGE cluster (at PSB), but can 
also run locally. IdentifyProtein.py which runs blastp (via Biopython) however does not currently function on the 
cluster, but does run locally. (More about this on the IdentifyProtein section below).



#### Directory structure & names:

While the root directory structure is hardwired into the software via a yaml file in the config folder /MutateCompute/config, 
the paths are built from the strings in the yaml via the Paths class (/MutateCompute/src/enums/) at runtime.

To change paths throughout the codebase to match your environment, just change them in the pathsAndDictionaries.yml file 
and the rest is automatically built at 'runtime' by set_up_paths(bool) which is in the Paths enum in Paths.py. 

The method takes one argument, a boolean, True to build paths using the cluster directory structure and names, 
False to build the local directory structure and names (according to root path in pathsAndDictionaries.yml).

The directory structure and names of directories are identical on the cluster except that instead of "MutateCompute", 
the root of the directory tree is called "SnpEffect".


To do this, copy all of /MutateCompute/src, /MutateCompute/bash and /MutateCompute/config into a folder in your user group on 
the cluster and name this SnpEffect. 
(For example, the absolute path of my cluster copy is /switchlab/group/shazib/SnpEffect). 

So, the Linux command (which is also works on MacOS) to copy the src folder on my computer is: 

scp -r -P 7788 /Users/u0120577/PycharmProjects/MutateCompute/input_data shazib@zeus.psb.ugent.be:/switchlab/group/shazib/SnpEffect

##
### DEPENDENCIES

The programs have a number of dependencies including Python 3. I use the the cluster's "module system" to install the latest 
Python3 that is installed on there. (Currently it is 3.5.1. Although the interpreter used to write the code is 3.7, the older 3
.5.1 version does not cause any major problems.). 
Use the virtual environment folder (venv) as it is currently in my cluster directory (/switchlab/group/shazib/SnpEffect/venv 
I've made a copy of it to /switchlab/group/tools). 
This is important because a few things were removed from the venv folder which is on Github as there seemed to be a couple of 
weird conflicts with 3.5.1 and 3.7 interpreter libraries. 


The PSB cluster Module System:
Load the Python module once with the 'autoload' command so that you don't have to keep loading it each time you log in:
module autoload Python 3.5.1 

(To see what is available to load use 'module available' )
 



Code execution is designed to be launched from one of the Python launcher scripts in /MutateCompute/src/launchers, via one of the
 bash scripts in /MutateCompute/bash/. Thus, to run FoldX programs, cd to the bash folder and use the command line argument:
 
sh KickOffZeus.sh

From terminal, cd to wherever you have downloaded the MutateCompute codebase. On my computer it looks like this:
cd /Users/u0120577/PycharmProjects/MutateCompute/bash/  

To run from Pycharm IDE either locally or on the cluster, select Run from the toolbar:
Run|Edit Configurations|Python|KickOffPy and select Apply/Run. 
The script path should be: /Users/u0120577/PycharmProjects/MutateCompute/src/launchers/KickOff.py
The working directory: /Users/u0120577/PycharmProjects/MutateCompute/src


The Python Classes & Scripts
--
####IdentifyProtein.py
<Only runs locally at the moment>
Located in the /src directory, this class (and Biopython class) automates identifying the 32431 homology model 
pdbs generated by collaborators (G.Vriend Radboudumc group).  
This class encapsulates methods for taking a protein sequence (FASTA), running blastp alignment tool on it against the 
SwissProt database and writing csv files of the protein(s) that give 100% identity hits. The csv files include the Accession 
Id, name, alternative names and positions of the hit.

IdentifyProtein uses the open-source Biopython tools (https://biopython.org/) to run blastp. I wrote a class to encapsulate this 
which I called Biopython (just to confuse things). This performs all interactions with the open-source Biopython tools. 
The blastp does not currently run on the cluster due to some conflict with the cluster's Biopython module that I did not manage
 to resolve. However, running IdentifyProtein & Biopython locally is currently set up to run on the NCBI servers anyway, so it 
 runs remotely in that way, just not through the PSB cluster. 
It is currently the slowest part of the the computations in MutateCompute.
 
####Agadir.py

Located in the /src directory, this class automates utilising any of the protein-aggregation and other algorithms
 included in the current Agadir wrapper written by J.Schymkowitz, F.Rousseau, L.Serrano and colleagues.

####FoldX.py

Located in the /src directory, this class automates utilising the protein structure-related force field 
computations of the FoldX suite of algorithms (written by J.Schymkowitz, F.Rousseau, L.Serrano and colleagues), currently set 
up and tested for FoldX version 3 and for the BuildModel and AnalyseComplex algorithms.

####MutateFasta.py

Located in the /src directory, this class automates taking protein sequences (FASTA) and generating FASTA files of every possible 
mutation at every residue. The input FASTA files were generated by collaborators (G.Vriend Radboudumc group). The outputs can 
be either a single text file (FASTA) per mutant sequence or a single file containing all possible mutant sequences for each input 
sequence. The latter is much better for disk-space and this has already been run for all 29611 FASTA sequences and is stored in
 subdirectories of 1000 files per directory in the lvs (large volume server) belonging to Switchlab, 
 smb://luna.kuleuven.be/lvs/GBW-0061_SWITCH/mutants_multifastas/29611_fastas_1000/ 

These files are taken as inputs to Agadir.py which writes a single FASTA file for each mutant, runs Agadir on it, then deletes 
the single extra file. 

